src/goto-instrument/reachability_slicer.cpp

```diff
@@ -20,41 +24,46 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "full_slicer_class.h"
 #include "reachability_slicer_class.h"
 
-void reachability_slicert::fixedpoint_assertions(
+/// Perform backwards breadth-first-search of the control-flow graph of the
+/// goto program, starting from the criterion. Set reaches_assertion to true
+/// for every instruction visited.
+/// \param criterion the criterion we are trying to hit
+/// \param is_threaded TODO
+void reachability_slicert::fixedpoint_to_assertions(
   const is_threadedt &is_threaded,
   slicing_criteriont &criterion)
 {
-  queuet queue;
-
-  for(cfgt::entry_mapt::iterator
-      e_it=cfg.entry_map.begin();
-      e_it!=cfg.entry_map.end();
-      e_it++)
-    if(criterion(e_it->first) ||
-       is_threaded(e_it->first))
-      queue.push(e_it->second);
-
-  while(!queue.empty())
-  {
-    cfgt::entryt e=queue.top();
-    cfgt::nodet &node=cfg[e];
-    queue.pop();
-
-    if(node.reaches_assertion)
-      continue;
-
-    node.reaches_assertion=true;
-
-    for(cfgt::edgest::const_iterator
-        p_it=node.in.begin();
-        p_it!=node.in.end();
-        p_it++)
-    {
-      queue.push(p_it->first);
-    }
-  }
+  std::vector<cfgt::node_indext> queue;
+  for(const auto &e_it : cfg.entry_map)
+    if(criterion(e_it.first) || is_threaded(e_it.first))
+      queue.push_back(e_it.second);
+
+  std::vector<cfgt::node_indext> reachable = cfg.get_reachable(queue, false);
+  for(const auto index : reachable)
+    cfg[index].reaches_assertion = true;
 }
 
+/// Perform forwards breadth-first-search of the control-flow graph of the
+/// goto program, starting from the criterion. Set reacheable_from_assertion
+/// to true for every instruction visited.
+/// \param criterion the criterion we are trying to hit
+/// \param is_threaded TODO
+void reachability_slicert::fixedpoint_from_assertions(
+  const is_threadedt &is_threaded,
+  slicing_criteriont &criterion)
+{
+  std::vector<cfgt::node_indext> queue;
+  for(const auto &e_it : cfg.entry_map)
+    if(criterion(e_it.first) || is_threaded(e_it.first))
+      queue.push_back(e_it.second);
+
+  std::vector<cfgt::node_indext> reachable = cfg.get_reachable(queue, true);
```

Comment:
use `const auto`

Reply:
Why `auto`? Type isn't repeated on rhs or particularly complicated. `const &` seems sensible though?

Reply:
The type name is long, and we don't use vector specific functions on this object, it's just used as a container, we could for instance change the return type of get_reachable to a set. This is maybe a personal preference  so I'm not insisting. 
I think making it a reference does not change anything since `get_reachable` creates a new vector, maybe a c++ expert could confirm that?

Reply:
The compiler will optimise away the copy on return in this case. Using a `const &` would be C++03 style, but is unnecessary in C++11.